src/std/misc/repr.ss

```diff
@@ -0,0 +1,129 @@
+;; Call the display-element function on each element.
+;; : <- (List Any) port: Port prefix: String separator: String suffix: String display-element: (<- Any)
+(def (display-separated
+      list
+      (port (current-output-port))
+      prefix: (prefix "")
+      separator: (separator " ")
+      suffix: (suffix "")
+      display-element: (display-element display))
+  (display prefix port)
+  (def first? #t)
+  (for-each! ;; NB: also works on improper lists.
+   list
+   (lambda (elem)
+     (if first?
+       (set! first? #f)
+       (display separator port))
+     (display-element elem port)))
+  (display suffix port))
+
+;; Print an evaluable representation of an object on the given port with the given options.
+;; The port defaults to (current-output-port). The options are reserved for future use.
+;; : <- Any (Optional Port) (Optional Table)
+(def (print-representation x (port (current-output-port)) (options (current-representation-options)))
+
+  ;; Our universal utilities: print (recurse), repr, display, write
+  (def (p y (q port)) (print-representation y q options))
+  (def (r y) (repr y options))
+  (def (d y) (display y port))
+  (def (w y) (write y port))
+  (def (simple? x) (or (number? x) (boolean? x) (string? x) (char? x) (void? x) (eof-object? x)))
+
+  (cond
+   ;; Simplest case: just write it.
+   ((simple? x)
+    (w x))
+   ((or (symbol? x) (null? x)) ;; requires slightly more care: write it after a quote.
+    (d "'") (w x))
+   ((pair? x) ;; pair: print as [ ... ].
+    (display-separated x port prefix: "[" separator: " " display-element: p)
+    (let ((end (cdr (last-pair x))))
+      (cond
+       ((null? end) (void))
+       ((simple? end) (d " . ") (p end))
+       (else (d " ") (p end) (d " ..."))))
+    (d "]"))
+   ((vector? x) ;; vector: print as (vector ...).
+    (display-separated (vector->list x) port
+                       prefix: "(vector " separator: " " display-element: p suffix: ")"))
+   ((u8vector? x) ;; u8vector: print as #u8(...), knowing that elements are all bytes.
+    (display-separated (u8vector->list x) port
+                       prefix: "#u8(" separator: " " display-element: d suffix: ")"))
+   ((hash-table? x) ;; hash-table: print as (hash ...)
+    ;; NB: also assumes (1) it is a equal? table, and (2) you use :std/sugar ...
+    (display-separated
+     (sort (map (lambda (k) (cons (r k) (hash-ref x k))) (hash-keys x)) ;; sort keys by repr
+           (lambda (krv1 krv2) (string<? (car krv1) (car krv2))))
+     port prefix: "(hash " suffix: ")"
+     display-element: (lambda (x _) (match x ([kr . v] (d "(") (d kr) (d " ") (p v) (d ")"))))))
+   ((and (procedure? x) (##procedure-name x))
+    => (lambda (name) (display name port)))
+   ((and (object? x) (find-method (object-type x) ':pr))
+    => (lambda (m) (m x port options)))
+   ((and (object? x) (assgetq transparent: (type-descriptor-plist (object-type x))))
+    (display-separated
+     (cdr (struct->list x)) port
```

Comment:
this has to discrimate between structs and classes, as the latter should be printed with slot keyword names.

Reply:
ie with class->list

Reply:
i'll merge and fix.